README.md
# svt-text
svt-text is an unofficial terminal viewer for SVT Text

## Example Usage
svt-text 100-105 202 300-303 401

## Installation
pip3 install svt-text

## Screenshots

![Page 100](/100.png?raw=true "Page 100")

![Page 401](/401.png?raw=true "Page 401")


## License

The source code is released under the AGPL license, but the tiles are not.


CHANGELOG.md
0.2.1 (2022-06-05)

* Faster startup speed.

0.2.0 (2022-06-04)

* Made the scraping work with the redesigned webpage.
* Use Unicode 13 box characters instead of Braille.
* Add config file to customize colors and more.
* Add caching.


codegen.py
import base64
import io
import os

import numpy as np
from PIL import Image

TWOP = np.array([1,2,4,8,16,32,64,128], dtype=np.uint8)

code = ['# generated code\nnames = [']
image_bytes = []
for fn in os.listdir('svt_text/tiles'):
    with open('svt_text/tiles/' + fn, 'rb') as f:
        fn = fn.replace('\'', '\\\'')
        code.append('  \'%s\',' % fn)

        content = f.read()
        f = io.BytesIO(content)
        im = Image.open(f)
        data = np.asarray(im.getdata())
        color_index_0 = data[0]
        for x in data:
            if x != color_index_0:
                color_index_1 = x
                break

        onebit = data == color_index_0
        key = tuple(np.dot(onebit.reshape(-1, 8), TWOP))
        image_bytes.append(bytes(key))

code.append(']')
code.append('files = %s\n' % base64.b64encode(b''.join(image_bytes)))
with open('svt_text/generated.py', 'w') as f:
    f.write('\n'.join(code))


text-tv.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""Display svt-text in a terminal.

Example usage:
  ./text-tv.py 100
"""

from html.parser import HTMLParser
from typing import List, Optional, Tuple
import re
import sys

try:
    import requests
except ModuleNotFoundError:
    print("'Requests' library is required")
    sys.exit(1)

classTable = {
    #'B'  : '\033[34m',
    #'B'  : '\033[94m',
    # Ubuntu blue is really weird, so specify 24bit color manually.
    # I don't feel great about this solution, may play bad with custom palettes.
    'B': '\033[38;2;0;0;255m',
    'R': '\033[91m',
    'G': '\033[92m',
    'Y': '\033[93m',
    'M': '\033[95m',
    'C': '\033[96m',
    'W': '\033[97m',
    'bgBK': '\033[40m',
    'bgR': '\033[41m',
    'bgB': '\033[44m',
    'bgG': '\033[102m',
    'bgY': '\033[103m',
    'bgC': '\033[106m',
    'bgW': '\033[107m',
    # Should be double height, but that's not happening. Just bold it.
    'DH': '\033[1m',
}


def getChar(image_number: int) -> str:
    """Converts an image number to a character.

    For instance if filename was 112.gif
    imageNumber is '112'
    Based on experiments the images have 2x3 pixels, one bit color depth.
    Valid filename numbers are 32-63 and 96-127

    By subtracting 32 from the numbers in the low range, and 64 from the numbers
    in the high range we get a number between 0 and 63. Every bit of that number
    corresponds to a pixel in the following pattern.
      1  2
      4  8
     16 32
    """

    if image_number >= 64:
        x = image_number - 64
    else:
        x = image_number - 32

    mode = 'braille'

    if mode == 'braille':
        # 6 dot braille uses the transpose bit pattern
        # 1  8
        # 2 16
        # 4 32
        # So shuffle all the bits around and add the braille offset
        x = ((x & (1+32)) | ((x & 2) << 2) | ((x & 4) >> 1)
                          | ((x & 8) << 1) | ((x & 16) >> 2))
        return chr(10240 + x)
    elif mode == 'boxele':
        # Unicode has 4-pixel box-elements.
        # Convert from 6 pixel to 4 pixel by throwing away middle row.
        x = (x & 3) + (x & 48) // 4
        # The box elements don't seem to have any particular order, so we list
        # them and index.
        return [' ', '▘', '▝', '▀', '▖', '▌', '▞', '▛', '▗', '▚',
                '▐', '▜', '▄', '▙', '▟', '█'][x]
    else:
        if x == '32':
            return ' '
        return '█'


def getEscapes(classes: List[str]) -> str:
    """Takes a list of classes, returns string with corresponding escapes"""
    ret = ''
    for c in classes:
        if c in classTable:
            ret += classTable[c]
    return ret


class SVTParser(HTMLParser):
    """Parses an svt-text page, and creates terminal text in self.s

    Takes html from svt-text page.  Converts fore- and background colors from
    styled spans to terminal escapes. Some spans have background images. They are
    converted to a text representation.
    """

    def __init__(self):
        super().__init__()
        self.s = ''

        # Keep track of classes of nested spans by pushing previous styles onto a
        # stack, so we can pop when we get a closing tag.
        self.stack = [[]]  # type: List[List[str]]
        # Keep track of nested background-images.
        self.bg = [None]  # type: List[Optional[int]]
        self.bg_pattern = re.compile(
            'background: url\(../../images/mos/(?:[A-Z])/(\d+).gif\)')
        self.in_page = False

    def span_enter(self, attrs: List[Tuple[str, str]]):
        new_classes = []  # type: List[str]
        number = None
        for attr in attrs:
            if attr[0] == 'class' and attr[1]:
                new_classes = attr[1].split(' ')
            if attr[0] == 'style':
                m = self.bg_pattern.match(attr[1])
                if m is not None:
                    number = int(m.group(1))
        self.s += getEscapes(new_classes)
        self.stack.append(self.stack[-1] + new_classes)
        self.bg.append(number)

    def span_exit(self):
        # Restore styles
        self.s += '\033[0m'
        self.stack.pop()
        self.bg.pop()
        self.s += getEscapes(self.stack[-1])

    def pre_enter(self, attrs: List[Tuple[str, str]]):
        assert not self.in_page
        if attrs == [('class', 'root')]:
            # First page
            self.in_page = True
        elif attrs == [('class', 'root sub')]:
            # Subsequent page
            self.s += '\n\n'
            self.in_page = True

    def pre_exit(self):
        if self.in_page:
            self.in_page = False

    def handle_starttag(self, tag: str, attrs: List[Tuple[str, str]]):
        if tag == 'pre':
            self.pre_enter(attrs)
        elif self.in_page and tag == 'span':
            self.span_enter(attrs)

    def handle_endtag(self, tag: str):
        if tag == 'pre':
            self.pre_exit()
        elif self.in_page and tag == 'span':
            self.span_exit()

    def handle_data(self, data: str):
        if not self.in_page:
            return
        cur_bg = self.bg[-1]
        if cur_bg is None:
            # We don't have any background image. Just put all data in output
            self.s += data
        else:
            # Replace spaces (that should be all characters) with a character
            # that represents the background image
            replacement = getChar(cur_bg)
            self.s += data.replace(' ', replacement)


def main():
    # Cast to int and back to string to validate the argument is really a number.
    n = int(sys.argv[1])
    response = requests.get('https://www.svt.se/svttext/tv/pages/'
                            + str(n) + '.html')
    if response.status_code != 200:
        friendly = ''
        if response.status_code == 404:
            friendly = ': File not found'
        print('Failed to fetch page. Got %d%s' %
              (response.status_code, friendly))
        sys.exit(1)

    parser = SVTParser()
    parser.feed(response.text)
    print(parser.s)


main()


svt_text/common.py
#  Copyright (C) 2022  Rickard Norlander
#
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU Affero General Public License version 3
#  as published by the Free Software Foundation.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU Affero General Public License for more details.
#
#  You should have received a copy of the GNU Affero General Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>.

import sys

__version__ = '0.2.1'

# These "constants" are set based on configuration and command line arguments.
VERBOSE = False
BRAILLE = False
CACHE = 1

# Actually constant
DH_FALSE = 0
DH_TRUE = 1
DH_TOP = 2
DH_BOTTOM = 3

class ParsedTile:
    """Data about a parsed tile."""
    def __init__(self, char, shape_id, fg_color, bg_color, dh_type):
        self.char = char
        self.shape_id = shape_id
        self.fg_color = fg_color
        self.bg_color = bg_color
        self.dh_type = dh_type

    def __repr__(self):
        return self.char or str(self.shape_id)

    def copy(self):
        return ParsedTile(self.char, self.shape_id, self.fg_color,
                          self.bg_color, self.dh_type)

    def is_shape_not_space(self):
        return self.shape_id > 0

    def is_shape_or_space(self):
        return self.char == ' ' or self.shape_id > 0

    def invert_shape(self, newbg=None):
        if newbg is None:
            newbg = self.fg_color
        self.fg_color, self.bg_color = self.bg_color, newbg
        self.shape_id = 63 - self.shape_id
        if self.shape_id == 0:
            self.char = ' '
        else:
            self.char = None

def err_print(msg: str):
    """Print a message to stderr"""
    print(msg, file=sys.stderr)

def err_verbose(msg: str):
    """Print a message to stderr if verbose flag is set."""
    if VERBOSE:
        print(msg, file=sys.stderr)


svt_text/render.py
"""Renders SVT Text in a terminal using braille characters.

Renders a page from a list of list of lists of ParsedTile.  Fore- and
background colors are set using terminal escape codes. Double width
text is rendered as bold. Can use either Braille characters or Unicode
13 2x3 box characters.
"""

#  Copyright (C) 2022  Rickard Norlander
#
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU Affero General Public License version 3
#  as published by the Free Software Foundation.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU Affero General Public License for more details.
#
#  You should have received a copy of the GNU Affero General Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>.

from typing import List

from svt_text import common

FG_ESCAPES = {
    'K': '\033[35m',
    'R': '\033[91m',
    'G': '\033[92m',
    'B': '\033[38;2;0;50;255m',
    'Y': '\033[93m',
    'M': '\033[95m',
    'C': '\033[96m',
    'W': '\033[97m'
}

BG_ESCAPES = {
    'K': '',
    'R': '\033[41m',
    'G': '\033[102m',
    'B': '\033[44m',
    'Y': '\033[103m',
    'M': '\033[105m',
    'C': '\033[106m',
    'W': '\033[107m',
}

BOLD_ESCAPE = '\033[1m'
RESET_ESCAPE = '\033[0m'

class Style:
    """Terminal character style."""
    def __init__(self, pt):
        self.fg_color = pt.fg_color
        self.bg_color = pt.bg_color
        self.bold = pt.dh_type == common.DH_TRUE

    def as_escapes(self, reset=True):
        return ((RESET_ESCAPE if reset else '') +
                (BG_ESCAPES[self.bg_color] if self.bg_color else '') +
                (FG_ESCAPES[self.fg_color] if self.fg_color else '') +
                (BOLD_ESCAPE if self.bold else ''))

    def __eq__(self, other):
        if other is None:
            return False
        return (self.fg_color == other.fg_color and
                self.bg_color == other.bg_color and
                self.bold == other.bold)

def getboxchar(x: int):
    """Get box character for shape x

    2x3 box characters use the same bit pattern as shape_id, namely

      1  2
      4  8
     16 32

    EXCEPT! That it has gaps corresponding to

    00  10  01  11
    00  10  01  11
    00  10  01  11

    These occur at 21 * n for n=0,1,2,3.  We can use space for the
    first and 2x2 box elements for the rest.
    """
    d, m = divmod(x + 20, 21)
    if m == 20:
        return ' ▌▐█'[d]
    return chr(0x1FB00 + x - d)

def getbraille(x: int):
    """Get braille character for shape x.

    6 dot braille uses the following bit pattern

    1  8
    2 16
    4 32

    Shuffle the bits around and apply the braille offset.
    """
    x = ((x & (1+32)) | ((x & 2) << 2) | ((x & 4) >> 1)
         | ((x & 8) << 1) | ((x & 16) >> 2))
    return chr(10240 + x)

def render_row(parsed_tiles: List[common.ParsedTile]):
    prev_style = None
    for pt in parsed_tiles:
        style = Style(pt)
        if style != prev_style:
            prev_style = style
            print(style.as_escapes(), end='')
        print(pt.char, end='')
    if prev_style is not None:
        print(RESET_ESCAPE, end='')
    print()

def fgbg_fixups(parsed_tiless: List[List[common.ParsedTile]], page: int, subpageid: int):
    """Apply foreground-background shifts.

    In the original page, a white shape on a black background is
    indistinguishable from a black shape on a white background. But
    this is not true when we render. Two reasons: Color and braille.

    The terminal background palette may not be the same as the
    foreground palette. If the rendering uses braille mode, then the
    second issue is that a braille dot will not fully cover the space
    - it will let some background show through. Sometimes a fully
    colored tile may even be rendered as a 6-dot braille on an
    inferred background, for consistency with the surroundings.

    The best looking foreground-background choice must be made. This
    is done heuristically with some manual overrides.
    """
    lastbg = 'K'
    for parsed_tiles in parsed_tiless:
        for parsed_tile in parsed_tiles:
            if not parsed_tile.is_shape_or_space():
                continue

            fg, bg = parsed_tile.fg_color, parsed_tile.bg_color
            if parsed_tile.char == ' ':
                if (bg, lastbg) == ('Y', 'B'):
                    parsed_tile.invert_shape(lastbg)
                elif 401 == page and bg != 'K':
                    parsed_tile.invert_shape('K')
                elif 777 == page and subpageid == 2 and (bg, fg) == ('C', 'K'):
                    parsed_tile.invert_shape('K')
                elif 777 == page and subpageid < 2:
                    lastbg = bg
                elif page not in [129, 149] and bg == 'W':
                    parsed_tile.invert_shape(lastbg)
                else:
                    lastbg = bg
            else:
                if fg == 'K' or bg == 'W':
                    parsed_tile.invert_shape()
                elif (bg, fg) == ('Y', 'B'):
                    parsed_tile.invert_shape()

def remove_fg_back_and_forth(parsed_tiless: List[List[common.ParsedTile]]):
    for parsed_tiles in parsed_tiless:
        for i in range(len(parsed_tiles)):
            if i < 2: continue

            if parsed_tiles[i-1].char != ' ': continue
            if parsed_tiles[i-2].char == ' ' or parsed_tiles[i].char == ' ': continue
            if parsed_tiles[i-2].fg_color != parsed_tiles[i].fg_color: continue
            parsed_tiles[i-1].fg_color = parsed_tiles[i].fg_color

def apply_shapes(parsed_tiless: List[List[common.ParsedTile]], braille: bool):
    for parsed_tiles in parsed_tiless:
        for parsed_tile in parsed_tiles:
            if parsed_tile.is_shape_not_space():
                if braille:
                    parsed_tile.char = getbraille(parsed_tile.shape_id)
                else:
                    parsed_tile.char = getboxchar(parsed_tile.shape_id)


def render_page(parsed_tilesss: List[List[List[common.ParsedTile]]], page: int):
    for subpageid, parsed_tiless in enumerate(parsed_tilesss):
        fgbg_fixups(parsed_tiless, page, subpageid)
        remove_fg_back_and_forth(parsed_tiless)
        apply_shapes(parsed_tiless, common.BRAILLE)
        for parsed_tiles in parsed_tiless:
            render_row(parsed_tiles)


svt_text/svt_text.py
"""Display svt-text in a terminal.

Example usage:
  svt-text 100
"""

#  Copyright (C) 2022  Rickard Norlander
#
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU Affero General Public License version 3
#  as published by the Free Software Foundation.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU Affero General Public License for more details.
#
#  You should have received a copy of the GNU Affero General Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>.

import argparse
import configparser
import sys
import os
from typing import List

import requests

from svt_text import common
from svt_text.common import err_print, err_verbose
from svt_text import fetch
from svt_text import render

arg_parser = argparse.ArgumentParser(
    description='Display svt-text in a terminal.', prog='svt-text',
    allow_abbrev=False, prefix_chars='-+')

class YesNo(argparse.Action):
    def __call__(self, parser, namespace, values, option):
        is_negated = option[0] == '+' or option[:5] == '--no-'
        setattr(namespace, self.dest, not is_negated)

def make_yes_no(name, help, short):
    if short is None:
        shorts = []
    else:
        shorts = ['-%s' % short, '+%s' % short]
    arg_parser.add_argument(*shorts, '--%s' % name, '--no-%s' % name,
                            dest=name, nargs=0, action=YesNo, help=help)

arg_parser.add_argument(
    '--cache', metavar='TIME', help='cache pages for N minutes. Inf to always cache')
make_yes_no('braille', 'draw with braille characters, needed if font lacks 2x3 box elements.', None)
arg_parser.add_argument(
    '--version', action='version', version='%(prog)s ' + common.__version__)
make_yes_no('verbose', 'show diagnostic information.', 'v')

def parse_page_range(page_range: str) -> List[int]:
    """Parses the page range command line argument."""
    try:
        splitted = page_range.split('-')
        if len(splitted) == 1:
            page = int(page_range)
            return [page, page]
        if len(splitted) == 2:
            return [int(x) for x in splitted]
        raise ValueError()
    except ValueError:
        # We can get here from either the int casts or the explicit raise.
        raise argparse.ArgumentTypeError(
            "'%s' is not a valid page range" % page_range)


arg_parser.add_argument(
    'page_range', type=parse_page_range, nargs='+',
    help='either N for a single page, or M-N for a range of pages')
args = argparse.Namespace()

def get_pages_to_fetch() -> List[int]:
    pages = []  # type: List[int]
    invalid_pages = False
    for page_range in args.page_range:
        if page_range[0] < 100:
            page_range[0] = 100
            invalid_pages = True
        if page_range[1] > 999:
            page_range[1] = 999
            invalid_pages = True
        pages.extend(range(page_range[0], page_range[1] + 1))

    if invalid_pages and not pages:
        err_print('All pages outside the valid range 100-999.')
        sys.exit(1)

    if invalid_pages and pages:
        err_print('There were pages outside of the valid range 100-999. '
                 'They will be skipped.')
    if not invalid_pages and not pages:
        err_print('No pages to fetch')
    return pages

def get_cache_dir():
    root_cache = os.getenv('XDG_CACHE_HOME')
    if root_cache is None:
        home = os.getenv('HOME', '/tmp')
        root_cache = os.path.join(home, '.cache')
    return os.path.join(root_cache, 'svt-text')

def get_config_paths():
    ret = []
    root_config = os.getenv('XDG_CONFIG_HOME')
    if root_config is None:
        home = os.getenv('HOME')
        if home is not None:
            root_config = os.path.join(home, '.config')
    if root_config is not None:
        ret.append(os.path.join(root_config, 'svt-text', 'config'))
    return ret

def main():
    global args
    args = arg_parser.parse_args()

    config = configparser.ConfigParser()
    config.read(get_config_paths())
    conf_valid = True
    try:
        configcache = config.get('Common', 'cache', fallback=common.CACHE)
        if configcache == 'Inf':
            common.CACHE = 'Inf'
        else:
            common.CACHE = int(configcache)
        common.VERBOSE = config.getboolean('Common', 'verbose', fallback=common.VERBOSE)
        common.BRAILLE = config.getboolean('Common', 'braille', fallback=common.BRAILLE)
    except ValueError:
        err_print('Invalid config file')
        conf_valid = False
    if conf_valid and 'Escapes' in config:
        for k, v in config['Escapes'].items():
            k = k.upper()
            v = v.encode('raw_unicode_escape').decode('unicode_escape')
            if k in render.FG_ESCAPES:
                render.FG_ESCAPES[k] = v
            elif len(k) == 3 and k[:2] == 'BG' and k[2] in render.BG_ESCAPES:
                render.BG_ESCAPES[k[2]] = v
            elif k == 'BOLD':
                render.BOLD_ESCAPE = v

    if args.verbose is not None:
        common.VERBOSE = args.verbose
    if args.braille is not None:
        common.BRAILLE = args.braille
    if args.cache  is not None:
        if args.cache == 'Inf':
            common.CACHE = 'Inf'
        else:
            try:
                common.CACHE = int(args.cache)
            except ValueError:
                err_print('Invalid cache duration')
                arg_parser.print_help()
                sys.exit(1)
    err_verbose('svt-text version %s' % common.__version__)
    cache_dir = get_cache_dir()
    os.makedirs(cache_dir, exist_ok=True)
    fetcher = fetch.Fetcher(None, cache_dir)
    last_page, next_page = None, None
    fetched_one = False
    showed_one = False
    pages = get_pages_to_fetch()

    err_verbose('Fetching %s pages' % len(pages))
    try:
        for page in pages:
            if next_page is not None and last_page < page < next_page:
                err_verbose('Skipping page %s' % page)
                continue
            next_page, page_info = fetcher.get_page(page)
            last_page = page
            fetched_one = True
            if len(page_info) == 0:
                err_verbose('Page %s was empty' % page)
                continue
            render.render_page(page_info, page)
            showed_one = True
    except requests.exceptions.ConnectionError as err:
        err_print('Fatal error: Failed to download %s' % err.request.url)
        sys.exit(1)
    except fetch.FetchException as err:
        err_print(err.args[0])
        sys.exit(1)

    if fetched_one and not showed_one:
        err_print('All pages were empty')

    sys.exit(0)

if __name__ == '__main__':
    main()


svt_text/fetch.py
"""Fetches and interprets SVT Teletext pages.

Fetches web pages corresponding to the teletext pages. A json data
structure is located. Inside is one or more base64-encoded gif images,
each containing a rendered sub page. The images are made up of tiles
of a fixed size where a tile could either be a character, a
decoration, or half a double-height character. The tiles in the image
are converted to monochrome and interpreted by comparing them to
reference tiles.

Example Usage:
  fetcher = Fetcher(None, None)
  fetcher.get_page(100)
"""

#  Copyright (C) 2022  Rickard Norlander
#
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU Affero General Public License version 3
#  as published by the Free Software Foundation.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU Affero General Public License for more details.
#
#  You should have received a copy of the GNU Affero General Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>.

import base64
import io
import json
import os
import re
import time
from typing import Dict, List, Optional, Tuple

import numpy as np
import PIL
from PIL import Image
import requests

from svt_text import common
import svt_text.generated
from svt_text.common import err_print, err_verbose, ParsedTile
from svt_text.common import DH_FALSE, DH_TRUE, DH_TOP, DH_BOTTOM

TWOP = np.array([1,2,4,8,16,32,64,128], dtype=np.uint8)
SCRIPT_PATTERN = re.compile(
    '<script id="__NEXT_DATA__" type="application/json">(.*)</script></body></html>')

COLOR_TABLE = {
    (0, 0, 0): 'K',
    (255, 0, 0): 'R',
    (0, 255, 0): 'G',
    (0, 0, 255): 'B',
    (255, 255, 0): 'Y',
    (255, 0, 255): 'M',
    (0, 255, 255): 'C',
    (255, 255, 255): 'W'
}
class Tile:
    """A tile from a page-image."""
    def __init__(self, data, key, colors):
        self.data = data
        self.key = key
        self.colors = colors

    @staticmethod
    def from_tile_im(tile_im: Image.Image):
        data = np.asarray(tile_im.getdata())
        color_index_0 = data[0]
        onebit = data == color_index_0
        key = tuple(np.dot(onebit.reshape(-1, 8), TWOP))
        return Tile(None, key, None)

    @staticmethod
    def _lookup_palette(palette: List[int], n: int):
        return (palette[3*n], palette[3*n+1], palette[3*n+2])

    def save(self, fn: str):
        im = Image.fromarray(self.data)
        return im.save(fn)


def merge_doubleheight(top: ParsedTile, bottom: ParsedTile):
    """Merges two halves of a double-height character."""
    assert top.bg_color == bottom.bg_color
    merged = bottom.copy()
    merged.dh_type = DH_TRUE
    if top.char == ' ' and bottom.char == ' ':
        return merged

    if top.char != ' ' and bottom.char != ' ':
        assert top.fg_color == bottom.fg_color

    assert top.is_shape_not_space() == bottom.is_shape_not_space()
    if bottom.is_shape_not_space():
        assert top.shape_id == bottom.shape_id
        return merged

    if top.char != ' ':
        assert top.dh_type == DH_TOP

    if bottom.char != ' ':
        assert bottom.dh_type == DH_BOTTOM

    if top.char == bottom.char:
        return merged

    sub_map = {'Fr': 'F', 'äa': 'ä', 'np': 'p', 'öo': 'ö', 'åa': 'å', 'il': 'i',
               'du': 'd', 'ug': 'y', 'hb': 'b', 'Hn': 'H', 'uv': 'v', 'cs': 's',
               'Yf': 'Y', 'ÖO': 'Ö', 'Tf': 'T', 'FL': 'E', 'Mn': 'M', '8U': '8',
               '35': '3', ' .': '.', ' -': '-', '6U': '6', 'ÄÅ': 'Ä', 'OQ': 'Q',
               'ée': 'é', ' ,': ',', '" ': '"'}
    c = sub_map.get('%s%s' % (top.char, bottom.char))
    if c is not None:
        if bottom.char == ' ':
            merged = top.copy()
            merged.dh_type = DH_TRUE
        merged.char = c
        return merged
    merged.char = '(%s,%s)' % (top.char, bottom.char)
    return merged

def merge_rows(parsed_tiless: List[List[ParsedTile]]):
    """Looks for double-height characters and merges their rows."""
    parsed_tiless2 = []
    for i, parsed_tiles in enumerate(parsed_tiless):
        num_top = sum(pt.dh_type == DH_TOP for pt in parsed_tiles)
        num_bottom = sum(pt.dh_type == DH_BOTTOM for pt in parsed_tiles)
        if num_top == 0 and num_bottom == 0:
            parsed_tiless2.append(parsed_tiles)
            continue
        assert num_bottom == 0 or num_top == 0
        if num_bottom == 0:
            continue
        assert i > 0
        last_pts = parsed_tiless[i-1]
        new_pts = []
        for last_pt, pt in zip(last_pts, parsed_tiles):
            new_pt = merge_doubleheight(last_pt, pt)
            new_pts.append(new_pt)
        parsed_tiless2.append(new_pts)
    return parsed_tiless2

class FetchException(Exception): pass

def index(values, keys):
    # Values is an array of shape[a][b][c]
    # Keys is an array of shape[a][b] where every value is an index into c
    values = values.reshape(-1, values.shape[-1])
    keys = keys.reshape(-1)
    return values[np.arange(keys.size), keys]

def get_tiles(im):
    pal = np.array(im.getpalette()).reshape(256, 3)
    char_palette = [COLOR_TABLE.get(tuple(row), 'K') for row in pal]
    char_palette = np.array(char_palette, dtype=object)

    imdata = np.array(im.getdata())

    width, height = im.size
    tile_width, tile_height = (13, 16)
    num_cols, num_rows = (40, 25)
    assert (width, height) == (tile_width * num_cols, tile_height * num_rows)

    tiles = imdata.reshape((num_rows, tile_height, num_cols, tile_width)).swapaxes(1, 2)
    binary_tiles = (tiles == tiles[:, :, 0:1, 0:1])

    color_0_inds = tiles[:, :, 0, 0]
    color_0 = char_palette[color_0_inds]

    linear_binary_tiles = binary_tiles.reshape(num_rows, num_cols, -1)
    first_different = np.argmin(linear_binary_tiles, axis=2)
    linear_tiles = tiles.reshape(num_rows, num_cols, -1)
    color_1_inds = index(linear_tiles, first_different).reshape(num_rows, num_cols)
    color_1 = char_palette[color_1_inds]

    keys = np.einsum('ijkm,m', binary_tiles.reshape(25, 40, 26, 8), TWOP)

    def make_tile(i, j):
        return Tile(data=binary_tiles[i,j], key=bytes(keys[i, j]),
                    colors=(color_0[i, j], color_1[i, j]))
    return [[make_tile(i, j) for j in range(num_cols)] for i in range(num_rows)]

class Fetcher():
    """Fetches and interprets SVT Teletext pages"""
    def __init__(self, extratiles_dir: Optional[str], cache_dir: Optional[str]):
        self.extratiles_dir = extratiles_dir
        self.cache_dir = cache_dir
        self.known_tiles, self.tile_data = self._read_tile_db()
        self.num_unknown_tiles: Dict[int, int] = {}

    def get_page(self, page_num: int) -> Tuple[int, List[List[List[ParsedTile]]]]:
        """Downloads and interprets a page.

        Returns a nested container of ParsedTile where the dimensions are
        [subpage][row][column]
        """
        response_text = self._download_html(page_num)
        next_page, subpages = self._parse_html(response_text)
        parsed_tilesss = []
        for subpage_num, subpage in enumerate(subpages):
            raw_tiles = get_tiles(subpage)
            parsed_tiless = self._scan_subpage(raw_tiles, page_num, subpage_num)
            parsed_tilesss.append(parsed_tiless)
        if page_num in self.num_unknown_tiles:
            err_print('Page %s has %s unknown tiles.' % (page_num, self.num_unknown_tiles[page_num]))
        return next_page, parsed_tilesss

    def _download_html(self, page: int) -> str:
        if self.cache_dir is not None:
            cache_path = os.path.join(self.cache_dir, str(page) + '.txt')
            try:
                age = (time.time() - os.path.getmtime(cache_path)) / 60
                if common.CACHE == 'Inf' or 0 <= age < common.CACHE:
                    with open(cache_path) as f:
                        ret = f.read()
                        err_verbose('Found page %s in cache' % page)
                        return ret
            except FileNotFoundError:
                pass
        response = requests.get('https://www.svt.se/text-tv/' + str(page))
        if response.status_code != 200:
            friendly = ''
            if response.status_code == 404:
                friendly = ': File not found'
            raise FetchException('Failed to fetch page. Got %d%s' %
                                   (response.status_code, friendly))
        if self.cache_dir is not None:
            with open(cache_path, 'w') as f:
                f.write(response.text)
        return response.text

    def _parse_html(self, response_text: str) -> Tuple[int, List[Image.Image]]:
        m = SCRIPT_PATTERN.search(response_text)
        if m is None:
            raise FetchException('Failed to parse, page had unexpected structure')
        try:
            o = json.loads(m.group(1))
            next_page = o["props"]["pageProps"]["nextPage"]
            if next_page:
                next_page = int(next_page)
            else:
                next_page = 1000

            images = []
            for subpage in o["props"]["pageProps"]["subPages"]:
                base64s = subpage["gifAsBase64"]
                f = io.BytesIO(base64.b64decode(base64s))
                im = Image.open(f)
                if im.size != (520, 400):
                    raise FetchException('Image had unexpected dimensions')
                images.append(im)
        except (KeyError, ValueError):
            raise FetchException('Failed to parse, page had unexpected structure')
        except PIL.UnidentifiedImageError:
            raise FetchException('Unable to decode image')
        return next_page, images

    def _scan_row(self, tile_row: List[Tile], page_num: int,
                  subpage_num: int, row_num: int) -> List[ParsedTile]:
        parsed_tiles = []
        for col_num, tile in enumerate(tile_row):
            if tile.key not in self.known_tiles:
                if self.extratiles_dir is not None:
                    tile.save(
                        "%s/%s-%s-%s-%s.pbm" % (self.extratiles_dir, page_num,
                                                subpage_num, row_num, col_num))
                self.known_tiles.add(tile.key)
            pt = self._tile_db_lookup(tile.key, page_num)
            pt.bg_color = tile.colors[0]
            pt.fg_color = tile.colors[1]
            parsed_tiles.append(pt)

        return parsed_tiles

    def _scan_subpage(self, subpage: List[List[Tile]], page_num: int, subpage_num: int):
        parsed_tiless = []
        for row_num, tile_row in enumerate(subpage):
            parsed_tiles = self._scan_row(tile_row, page_num, subpage_num, row_num)
            parsed_tiless.append(parsed_tiles)
        parsed_tiless = merge_rows(parsed_tiless)
        return parsed_tiless

    def _tile_db_lookup(self, key: Tuple[int, ...], page_num: int):
        tile_data = self.tile_data.get(key)
        if tile_data is None:
            self.num_unknown_tiles[page_num] = self.num_unknown_tiles.get(page_num, 0) + 1
            return ParsedTile(' ', 0 , 'K', 'K', DH_FALSE)
        return tile_data.copy()

    def _read_tile_db(self):
        known_tiles = set()
        tile_data = {}

        def handle_one(fn: str, im: Image.Image):
            if len(fn) < 5:
                return
            tile = Tile.from_tile_im(im)
            if tile.key in known_tiles:
                err_verbose('%s is dupe' % fn)
                return
            known_tiles.add(tile.key)
            if fn[:5] == 'slash':
                tile_data[tile.key] = ParsedTile('/', 0, None, None, DH_FALSE)
            if fn[:5] == 'slasT':
                tile_data[tile.key] = ParsedTile('/', 0, None, None, DH_TOP)
            if fn[:5] == 'slasB':
                tile_data[tile.key] = ParsedTile('/', 0, None, None, DH_BOTTOM)
            if fn[:4] == 'char':
                tile_data[tile.key] = ParsedTile(fn[4], 0, None, None, DH_FALSE)
            if fn[:4] == 'topp':
                tile_data[tile.key] = ParsedTile(fn[4], 0, None, None, DH_TOP)
            if fn[:4] == 'bott':
                tile_data[tile.key] = ParsedTile(fn[4], 0, None, None, DH_BOTTOM)
            if fn[:5] == 'shape':
                tile_data[tile.key] = ParsedTile(None, int(fn[5:-4]), None, None, DH_FALSE)

        for fn, encf in svt_text.generated.files.items():
            f = io.BytesIO(base64.b64decode(encf))
            im = Image.open(f)
            handle_one(fn, im)

        if self.extratiles_dir is not None:
            for fn in os.listdir(self.extratiles_dir):
                im = Image.open(os.path.join(self.extratiles_dir, fn))
                handle_one(fn, im)

        if len(tile_data) == 0:
            err_print('No tiles loaded')
        return known_tiles, tile_data
